### PR TITLE
Implement Sieve of Eratosthenes in prime.h

### PR DESCRIPTION
**Pull request by umutoztunc**

### CODE REVIEW

**Comment by umutoztunc** [comment]
On a second thought, I started to feel like class is not necessary for this. What about having a function with the following signature?
```cpp
vector<bool> sieve(const T &limit);
```

File: number_theory/prime.h [hunk 1]
```diff
@@ -17,13 +17,59 @@
 namespace tql {
 namespace number_theory {
 
+// Sieve of Eratosthenes.
+// It decides whether a number is prime or not for the numbers up to given
+// (inclusive) limit.
```
Comment (liangjs) [inline number_theory/prime.h:22]: to *a* given (inclusive) limit

File: number_theory/prime.h [hunk 2]
```diff
@@ -17,13 +17,59 @@
...
+template <typename T>
+class Sieve {
+  static_assert(std::numeric_limits<T>::is_integer,
+                "Sieve must use integer types.");
+
+ public:
+  // The type of numbers used by the Sieve.
+  using type = T;
+
+  explicit Sieve(const T &num_limit)
```
Comment (liangjs) [inline number_theory/prime.h:32]: Shall we have comments about its time complexity?

**Review by liangjs — DISMISSED** [review]
Looks nice.
Will we add unit test in this PR or later?

**Comment by liangjs** [comment]
> On a second thought, I started to feel like class is not necessary for this. What about having a function with the following signature?
> 
> ```c++
> vector<bool> sieve(const T &limit);
> ```

We can have such a function for users' sake. But I suggest we keep the Sieve class as well, so the code is consistent with other complex sieve classes.

File: number_theory/prime.h (hunk 2 above)
Comment (umutoztunc) [reply number_theory/prime.h:32]: I thought of it, but in my opinion we should focus on what the given code does. We can have a nice documentation where we also mention complexities, show examples, mention caveats etc.

**Comment by umutoztunc** [comment]
I am also unsure about defining a free function version of class APIs. Let's ignore that for now.

I have added the unit tests and made some slight adjustments. PTAL.

**Review by liangjs — APPROVED** [review]
LGTM.

**Pull request merged**